Mathlib/RingTheory/Flat/Basic.lean

```diff
@@ -499,6 +481,75 @@ theorem nontrivial_of_linearMap_injective_of_flat_right (f : R →ₗ[R] M) (h :
+
+/-- Tensor product of injective maps are injective under some flatness conditions.
+Also see `TensorProduct.map_injective_of_flat_flat'` and
+`TensorProduct.map_injective_of_flat_flat_of_isDomain` for different flatness conditions. -/
+lemma map_injective_of_flat_flat
+    (f : P →ₗ[R] M) (g : Q →ₗ[R] N) [Module.Flat R M] [Module.Flat R Q]
+    (hf : Function.Injective f) (hg : Function.Injective g) :
+    Function.Injective (TensorProduct.map f g) := by
+  rw [← LinearMap.lTensor_comp_rTensor]
+  exact (Module.Flat.lTensor_preserves_injective_linearMap g hg).comp
+    (Module.Flat.rTensor_preserves_injective_linearMap f hf)
+
+/-- Tensor product of injective maps are injective under some flatness conditions.
+Also see `TensorProduct.map_injective_of_flat_flat` and
+`TensorProduct.map_injective_of_flat_flat_of_isDomain` for different flatness conditions. -/
+lemma map_injective_of_flat_flat'
+    (f : P →ₗ[R] M) (g : Q →ₗ[R] N) [Module.Flat R P] [Module.Flat R N]
+    (hf : Function.Injective f) (hg : Function.Injective g) :
+    Function.Injective (TensorProduct.map f g) := by
+  rw [← LinearMap.rTensor_comp_lTensor]
+  exact (Module.Flat.rTensor_preserves_injective_linearMap f hf).comp
+    (Module.Flat.lTensor_preserves_injective_linearMap g hg)
+
+/-- Tensor product of linearly independent families is linearly
+independent under some flatness conditions.
+
+The flatness condition could be removed over domains.
+See `LinearIndependent.tensorProduct_of_isDomain`. -/
+lemma _root_.LinearIndependent.tensorProduct_of_flat_left [Module.Flat R M]
+    {ι ι' : Type*} {v : ι → M} (hv : LinearIndependent R v)
+    {w : ι' → N} (hw : LinearIndependent R w) :
+    LinearIndependent R fun i : ι × ι' ↦ v i.1 ⊗ₜ[R] w i.2 := by
+  rw [LinearIndependent]
+  convert (TensorProduct.map_injective_of_flat_flat _ _ hv hw).comp
+    (finsuppTensorFinsupp' _ _ _).symm.injective
+  rw [← LinearEquiv.coe_toLinearMap, ← LinearMap.coe_comp]
+  congr!
+  ext i
+  simp [finsuppTensorFinsupp'_symm_single_eq_single_one_tmul]
+
+/-- Tensor product of linearly independent families is linearly
+independent under some flatness conditions.
+
+The flatness condition could be removed over domains.
+See `LinearIndependent.tensorProduct_of_isDomain`. -/
+lemma _root_.LinearIndependent.tensorProduct_of_flat_right [Module.Flat R N]
+    {ι ι' : Type*} {v : ι → M} (hv : LinearIndependent R v)
+    {w : ι' → N} (hw : LinearIndependent R w) :
+    LinearIndependent R fun i : ι × ι' ↦ v i.1 ⊗ₜ[R] w i.2 :=
+  (((TensorProduct.comm R N M).toLinearMap.linearIndependent_iff_of_injOn
+    (TensorProduct.comm R N M).injective.injOn).mpr
+      (hw.tensorProduct_of_flat_left hv)).comp Prod.swap Prod.swap_bijective.injective
+
+variable (p : Submodule R M) (q : Submodule R N)
+
+/-- If p and q are submodules of M and N respectively, and M and q are flat,
+then `p ⊗ q → M ⊗ N` is injective. -/
+theorem _root_.Module.Flat.tensorProduct_mapIncl_injective_of_right
```

Comment:
```suggestion
theorem _root_.Module.Flat.tensorProductMapIncl_injective_of_right
```
same below

Reply:
This is just moved. Do you think I should rename it here? If so I would prefer to just drop `tensorProduct`.

Reply:
Ah sorry, if it's just moved, please ignore